### ISSUE TITLE
Sik-Ho Tang | Review -- BEiT: BERT Pre-Training of Image Transformers.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#137]
Sik-Ho Tang. [Review — BEiT: BERT Pre-Training of Image Transformers](https://sh-tsang.medium.com/review-beit-bert-pre-training-of-image-transformers-c14a7ef7e295).

**Comment by NorbertZheng** [comment]
## Overview
BEiT: BERT Pre-Training of Image Transformers, BEiT, by Microsoft Research, 2022 ICLR, Over 300 Citations.

 - Bidirectional Encoder representation from Image Transformers (BEiT) is proposed, where a masked image modeling (MIM) task to pretrain [Vision Transformer](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)s.
 - BEiT first “tokenizes” the original image into visual tokens. Then some image patches are randomly masked and fed into the backbone [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1).
 - The pre-training objective is to recover the original visual tokens based on the corrupted image.

**Comment by NorbertZheng** [comment]
## BEiT Architecture
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/e7b0be8b-87f6-4680-b06b-d28bd8168451)

### Overall Approach
 - Inspired by [BERT](https://sh-tsang.medium.com/59b1684882db), a pre-training task is proposed, namely, masked image modeling (MIM).
 - MIM uses two views for each images, i.e., image patches, and visual tokens.
 - The image is split into a grid of patches that are the input representation of backbone [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1).
 - The image is “tokenized” to discrete visual tokens by the latent codes of discrete VAE, where discrete VAE is from [DALL·E](https://sh-tsang.medium.com/review-dall-e-zero-shot-text-to-image-generation-f9de7a383374).

During pre-training, some proportion of image patches are randomly masked, and fed the corrupted input to [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1). The model learns to recover the visual tokens of the original image, instead of the raw pixels of masked patches.

**Comment by NorbertZheng** [comment]
### Image Representation
The images have two views of representations, namely, image patch, and visual tokens. The two types serve as input and output representations during pre-training, respectively.

#### Image Patches
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/3e79dd4d-9a84-4d38-93bd-b3067c24b16b)
Image Patches (Cut from the first figure).

 - The 2D image of the size $H\times W\times C$ is split into a sequence of patches $x_{p}$ ($p$ is from $1$ to $N$) of the size $P^{2}$, with the number of patch $N=\frac{HW}{P^{2}}$ patches.
 - The image patches $x_{p}$ are flattened into vectors and are linearly projected which is similar to word embeddings in [BERT](https://sh-tsang.medium.com/59b1684882db).

Particularly, BEiT splits each $224\times 224$ image into a $14\times 14$ grid of image patches, where each patch is $16\times 16$.

#### Visual Tokens
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/0b3b9b6f-4af3-4783-9113-e1f686e7fb93)
Visual Tokens (Cut from the first figure).

 - The image is represented as a sequence of discrete tokens obtained by an “image tokenizer”, instead of raw pixels.

Specifically, the image of the size $H\times W\times C$ is tokenized into $z=[z_{1},…,z_{N}]$, where the vocabulary $V=\\{1,…,|V|\\}$ contains discrete token indices.

 - The image tokenizer learned by discrete variational autoencoder (dVAE), by [DALL·E](https://sh-tsang.medium.com/review-dall-e-zero-shot-text-to-image-generation-f9de7a383374), is directly used.
 - There are two modules during visual token learning, namely, tokenizer and decoder.
 - The tokenizer $q(z|x)$ maps image pixels $x$ into discrete tokens $z$ according to a visual codebook (i.e., vocabulary).
 - The decoder $p(x|z)$ learns to reconstruct the input image $x$ based on the visual tokens $z$.
 - The vocabulary size is set to $|V| = 8192$.

**Comment by NorbertZheng** [comment]
### [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0) Backbone
 - Following [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0), the [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) backbone network is used.
 - The input of [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) is a sequence of image patches $x_{i}^{p}$.
 - The patches are then linearly projected to obtain patch embeddings $Ex_{i}^{p}$.
 - The standard learnable 1D position embeddings $E_{pos}$ are added to patch embeddings:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/1eabb0ee-7d09-42f6-a105-b35e456e2279)
 - The encoder contains L layers of [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) blocks:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/83f44f3e-1ed1-4f0b-b057-6e24c85f5c3e)
 - The output vectors of the last layer is:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/47e5806e-657e-475b-8997-793b7ab49e1d)

which are used as the encoded representations for the image patches, where $h_{i}^{L}$ is the vector of the $i$-th image patch.

 - [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)-Base is used, which is a 12-layer [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1) with 768 hidden size, and 12 attention heads. The intermediate size of feed-forward networks is 3072.

**Comment by NorbertZheng** [comment]
## BEiT Pretraining: Masked Image Modeling (MIM)
### Masked Image Modeling (MIM)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/a2cc71a1-fa45-45ac-bc72-6e5c24fcb55e)
BEiT Masked Image Modeling (MIM) (Cut from the first figure).

 - After spliting the image into image patches, as described above, approximately 40% image patches are randomly masked, where the masked positions are denoted as $M$. The masked patches are replaced with a learnable embedding $e[M]$. In BEiT, at most 75 patches are masked.
 - Then, the good and masked image patches are input into the L-layer [Transformer](https://sh-tsang.medium.com/review-attention-is-all-you-need-transformer-96c787ecdec1).
 - A softmax classifier is used to predict the corresponding visual tokens:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/72410e2e-b1dd-4f53-8c69-74048b33c452)

The pre-training objective is to maximize the log-likelihood of the correct visual tokens $z_{i}$ given the corrupted image:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/276355d5-a783-4c88-bf5e-721abdf5942e)

 - BEiT is pretrained on the training set of ImageNet-1K.
 - The pre-training runs for about 500k steps (i.e., 800 epochs) with 2k batch size. The 500k training steps take about five days using 16 Nvidia Tesla V100 32GB GPU cards.

**Comment by NorbertZheng** [comment]
###  Blockwise Masking
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/e5b282df-71d2-4452-8ea9-4d39746dbce4)
Blockwise Masking.

Blocks of patches are masked randomly as shown in the figure and algorithm above, instead of masking each patch individually in a random manner.

**Comment by NorbertZheng** [comment]
### From VAE Perspective
The BEiT pre-training can be viewed as variational autoencoder training:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/aa081ea3-f7b2-418a-ad24-292a11748e22)

 - In the first stage, the image tokenizer is obtained as a discrete variational autoencoder. Specifically, the first stage minimizes the reconstruction loss, with an uniform prior.
 - In the second stage, the prior $p_{\theta}$ is learnt while keeping $q_{\phi}$ and $p_{\psi}$ fixed.

Thus, the above equation is re-written as:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/cac930b9-44be-4b3e-a750-851b5cc31f1d)

where the second term is the proposed BEiT pre-training objective.

**Comment by NorbertZheng** [comment]
## Experimental Results
### ImageNet-1K & ImageNet-22K Pretraining, Image Classification on ImageNet-1K
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/b8b6113e-2fa3-4847-85c0-2e8fb0e849f6)
Top-1 accuracy on ImageNet-1K using full fine-tuning.

 - A simple linear classifier is employed as the task layer. Average pooling is used to aggregate the representations, and the global is fed to a softmax classifier.
 - Pre-trained BEiT significantly improves performance on both datasets.

BEiT improves the performance on ImageNet, which shows the effectiveness under the rich-resource setting.

 - Higher resolution improves the BEiT results by 1+ points on ImageNet.

More importantly, BEiT-384 pretrained on ImageNet-1K even outperforms supervised pre-training [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)-384 that uses ImageNet-22K, when they use the same input resolution.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/276910dc-8eb4-43c7-955d-1a3c8499f072)
Convergence curves of training [DeiT](https://sh-tsang.medium.com/review-deit-data-efficient-image-transformer-b5b6ee5357d0) from scratch and fine-tuning BEiT on ImageNet-1K.

Fine-tuning BEiT not only achieves better performance, but also converging much faster than training [DeiT](https://sh-tsang.medium.com/review-deit-data-efficient-image-transformer-b5b6ee5357d0) from scratch.

**Comment by NorbertZheng** [comment]
### Semantic Segmentation on [ADE20K](https://sh-tsang.medium.com/review-ade20k-semantic-understanding-of-scenes-through-the-ade20k-dataset-semantic-3414122f3cd3)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/d14dcfe2-6088-4a6e-b56f-5c762261950c)
Results of semantic segmentation on [ADE20K](https://sh-tsang.medium.com/review-ade20k-semantic-understanding-of-scenes-through-the-ade20k-dataset-semantic-3414122f3cd3).

 - The task layer used in SETR-PUP (Zheng et al., 2020), is used.
 - To be specific, the pretrained BEiT is used as a backbone encoder, and several deconvolution layers are incorporated as decoder to produce segmentation.

BEiT achieves better performance than supervised pretraining, although BEiT does not require manual annotations for pre-training.

 - Intermediate fine-tuning is performed for BEiT on ImageNet, i.e., first pretrained BEiT is fine-tuned on ImageNet, and then the model is fine-tuned on [ADE20K](https://sh-tsang.medium.com/review-ade20k-semantic-understanding-of-scenes-through-the-ade20k-dataset-semantic-3414122f3cd3).

Intermediate fine-tuning further improves BEiT on semantic segmentation.

**Comment by NorbertZheng** [comment]
### Ablation Study
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/78d7584e-39d9-4524-badd-fb59e37a9a61)
Ablation studies for BEiT pre-training on image classification and semantic segmentation.

 - Blockwise masking is beneficial on both tasks, especially on semantic segmentation.
 - The proposed masked image modeling (MIM) task significantly outperforms naïve pixel-level auto-encoding. The results indicate that **the prediction of visual tokens is the key ingredient of BEiT**.
 - Recovering all the visual tokens harms performance on downstream tasks.
 - **Pre-training the model longer (800 epochs) can further improve performance on downstream tasks.**

**Comment by NorbertZheng** [comment]
### Analysis on Self-Attention Map
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/62df0e28-2aa5-449c-952e-47a9c424fbaa)
Self-attention map for different reference points.

 - **The self-attention mechanism in BEiT can separate objects.**

After pre-training, BEiT learns to distinguish semantic regions using self-attention heads, without any task-specific supervision. Such knowledge acquired by BEiT potentially improves the generalization ability of fine-tuned models, especially on small-scale datasets.

**Comment by NorbertZheng** [comment]
## Further Results Using LayerScale in CaiT and Relative Position in Shaw NAACL’18 (Paper Appendix)
### Effects of LayerScale in [CaiT](https://sh-tsang.medium.com/review-cait-going-deeper-with-image-transformers-e05ffaea16f4) & Relative Position in [Shaw NAACL’18](https://sh-tsang.medium.com/review-self-attention-with-relative-position-representations-266ab2f78dd7)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/d3edb2bf-fd3b-4b2c-8182-4e7fc11bbbea)
Ablation studies of architecture variants on image classification and semantic segmentation.

LayerScale in [CaiT](https://sh-tsang.medium.com/review-cait-going-deeper-with-image-transformers-e05ffaea16f4), and relative position bias in [Shaw NAACL’18](https://sh-tsang.medium.com/review-self-attention-with-relative-position-representations-266ab2f78dd7), improve performance on ImageNet classification and [ADE20K](https://sh-tsang.medium.com/review-ade20k-semantic-understanding-of-scenes-through-the-ade20k-dataset-semantic-3414122f3cd3) semantic segmentation.

**Comment by NorbertZheng** [comment]
### ImageNet
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/8cf19f73-afa9-4be8-aff2-9b937a1c6d48)
Top-1 accuracy on ImageNet-1K fine-tuning.

BEiT-L fine-tuned on ImageNet-22K achieves comparable performance with [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0)-L trained on Google JFT-3B.

**Comment by NorbertZheng** [comment]
### [ADE20K](https://sh-tsang.medium.com/review-ade20k-semantic-understanding-of-scenes-through-the-ade20k-dataset-semantic-3414122f3cd3)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/c635b2b3-9a3f-4fb9-a3ad-b9ba55130d2f)
Performance comparison on the [ADE20K](https://sh-tsang.medium.com/review-ade20k-semantic-understanding-of-scenes-through-the-ade20k-dataset-semantic-3414122f3cd3) semantic segmentation.

The BEiT-L model obtains state-of-the-art performance on [ADE20K](https://sh-tsang.medium.com/review-ade20k-semantic-understanding-of-scenes-through-the-ade20k-dataset-semantic-3414122f3cd3), outperforms [Swin Transformer](https://sh-tsang.medium.com/review-swin-transformer-3438ea335585).

**Comment by NorbertZheng** [comment]
[DINO](https://sh-tsang.medium.com/cfddbb4d3549) applies self-supervised learning on [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0) using similar idea as [BYOL](https://sh-tsang.medium.com/review-byol-bootstrap-your-own-latent-a-new-approach-to-self-supervised-learning-6f770a624441). BEiT even uses the [BERT](https://sh-tsang.medium.com/59b1684882db) pretraining concept to have self-supervised learning on [ViT](https://sh-tsang.medium.com/review-vision-transformer-vit-406568603de0).

**Comment by NorbertZheng** [comment]
## Reference
 - [2022 ICLR] [BEiT] [BEiT: BERT Pre-Training of Image Transformers](https://arxiv.org/abs/2106.08254).
